sway/input/keyboard.c

```diff
@@ -485,6 +485,40 @@ static void handle_key_event(struct sway_keyboard *keyboard,
 	}
 
 	if (!handled || event->state == WLR_KEY_RELEASED) {
+		if (seat->im_relay.input_method) {
+			struct wlr_input_method_keyboard_grab_v2 *kb_grab =
+				seat->im_relay.input_method->keyboard_grab;
+			struct wlr_virtual_keyboard_v1 *virtual_keyboard =
+				wlr_input_device_get_virtual_keyboard(wlr_device);
+
+			// If event is from virtual keyboard of the same client as grab,
+			// do not send it back. TODO see swaywm/wlroots#2322
+			if (kb_grab && !(virtual_keyboard &&
+					wl_resource_get_client(virtual_keyboard->resource) ==
+					wl_resource_get_client(kb_grab->resource))) {
+				// Do not send release event to grab if the press event was not
+				// sent to grab.
+				if (event->state == WLR_KEY_RELEASED) {
```

Comment:
Hmm, I'm not quite sure I understand this part. Doesn't this mean we only forward RELEASED events to regular clients during a input-method grab?

Reply:
It was to send RELEASED to regular client instead of keyboard grab if its corresponding PRESSED was sent to regular client (pressed before grab happened).